### PR TITLE
Lock tree races

### PR DESCRIPTION
**Pull request by vlad-lesin**
TDB-3 : Lock tree stalls

See the initial pull request:

https://github.com/percona/PerconaFT/pull/360

Testing:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1889/
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/927/

### CODE REVIEW

**Comment by vlad-lesin** [comment]
@georgelorchpercona , this fix does not depend on "kill query" feature, could you please review it and consider it separately.

File: locktree/lock_request.cc [hunk 1]
```diff
@@ -290,47 +305,61 @@ int lock_request::retry(void) {
...
-    // it is important to note that this bit only provides an optimization.
-    // it is not problematic for it to be true when it should be false,
-    // but it can be problematic for it to be false when it should be true.
-    // therefore, the lock request code must ensures that when lock requests
-    // are added to this locktree, the bit is set.
-    // see lock_request::insert_into_lock_requests()
-    if (!info->should_retry_lock_requests) {
+    // if there are no pending lock requests than there is nothing to do
+    // the unlocked data race on pending_is_empty is OK since lock requests
+    // are retried after added to the pending set.
```
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: Comment formatting, line too long?

File: locktree/lock_request.cc [hunk 2]
```diff
@@ -290,47 +305,61 @@ int lock_request::retry(void) {
...
         return;
+
+    // get my retry generation (post increment of retry_want)
+    unsigned long long my_retry_want = (info->retry_want += 1);
+
+    toku_mutex_lock(&info->retry_mutex);
+
+    // here is the group retry algorithm.
+    // get the latest retry_want count and use it as the generation number of this retry operation.
+    // if this retry generation is > the last retry generation, then do the lock retries.  otherwise,
```
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: Comment formatting, lines too long?

File: locktree/lock_request.cc [hunk 3]
```diff
@@ -290,47 +305,61 @@ int lock_request::retry(void) {
...
+    // no lock retries are needed.
+    if ((my_retry_want - 1) == info->retry_done) {
+        for(;;) {
+            if (!info->running_retry) {
+                info->running_retry = true;
+                info->retry_done = info->retry_want;
+                toku_mutex_unlock(&info->retry_mutex);
+                retry_all_lock_requests_info(info);
+                if (after_retry_all_test_callback) after_retry_all_test_callback();
```
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: newline, line too long

File: locktree/lock_request.cc [hunk 4]
```diff
@@ -290,47 +305,61 @@ int lock_request::retry(void) {
...
+                break;
+            } else {
+                toku_cond_wait(&info->retry_cv, &info->retry_mutex);
+            }
+         }
     }
+    toku_mutex_unlock(&info->retry_mutex);
+}
 
+void lock_request::retry_all_lock_requests_info(lt_lock_request_info *info) {
```
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: Line too long

File: locktree/lock_request.cc [hunk 5]
```diff
@@ -395,6 +420,10 @@ void lock_request::set_start_test_callback(void (*f)(void)) {
     m_start_test_callback = f;
 }
 
+void lock_request::set_start_before_pending_test_callback(void (*f)(void)) {
```
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: Line too long

File: locktree/lock_request.h [hunk 6]
```diff
@@ -109,12 +109,19 @@ class lock_request {
     // effect: Retries all of the lock requests for the given locktree.
     //         Any lock requests successfully restarted is completed and woken up.
     //         The rest remain pending.
-    static void retry_all_lock_requests(locktree *lt);
+    static void retry_all_lock_requests(locktree *lt, void (*after_retry_test_callback)(void) = nullptr);
+    static void retry_all_lock_requests_info(lt_lock_request_info *info);
```
Comment (georgelorchpercona) [inline locktree/lock_request.h]: Lines too long

File: locktree/locktree.cc [hunk 7]
```diff
@@ -80,21 +80,23 @@ void locktree::create(locktree_manager *mgr, DICTIONARY_ID dict_id, const compar
...
+    ZERO_STRUCT(mutex);
+    toku_mutex_init(&mutex, nullptr);
+    retry_want = retry_done = 0;
+    ZERO_STRUCT(counters);
+    ZERO_STRUCT(retry_mutex);
+    toku_mutex_init(&retry_mutex, nullptr);
+    toku_cond_init(&retry_cv, nullptr);
+    running_retry = false;
+
+    TOKU_VALGRIND_HG_DISABLE_CHECKING(&pending_is_empty, sizeof(pending_is_empty));
```
Comment (georgelorchpercona) [inline locktree/locktree.cc]: Line too long

File: locktree/tests/lock_request_start_release_wait.cc [hunk 8]
```diff
@@ -0,0 +1,89 @@
+/* -*- mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+// vim: ft=cpp:expandtab:ts=8:sw=4:softtabstop=4:
+
+// test the race between start, release, and wait.  since start does not put its
+// lock request into the pending set, the blocking txn could release its lock before
+// the first txn waits.  this will block the first txn because its lock request is
+// not known when the lock is released.  the bug fix is to try again when lock retries
+// are locked out.
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_release_wait.cc]: Comment lines too long.

File: locktree/tests/lock_request_start_release_wait.cc [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+static int my_killed_callback(void) {
+    uint64_t t_now = toku_current_time_microsec();
+    assert(t_now >= t_wait);
+    if (t_now - t_wait >= my_killed_time*1000)
+        abort();
+    return 0;
+}
+
+static void locktree_release_lock(locktree *lt, TXNID txn_id, const DBT *left, const DBT *right) {
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_release_wait.cc]: Line too long

File: locktree/tests/lock_request_start_retry_race.cc [hunk 10]
```diff
@@ -97,20 +89,13 @@ static void run_locker(locktree *lt, TXNID txnid, const DBT *key) {
 
         toku_pthread_yield();
         if ((i % 10) == 0)
-            std::cout << toku_pthread_self() << " " << i << std::endl;
+            std::cout << std::this_thread::get_id() << " " << i << std::endl;
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race.cc]: Line too long and prefer to not use std::iostreams, printf to stderr instead

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+// Suppose that 3 threads are running a lock acquire, release, retry sequence.  There is
+// a race in the retry algorithm with 2 threads running lock retry simultaneously.  The
+// first thread to run retry sets a flag that will cause the second thread to skip the
+// lock retries.  If the first thread progressed past the contended lock, then the second
+// threa will HANG until its lock timer pops, even when the contended lock is no longer held.
+
+// This test exposes this problem as a test hang.  The group retry algorithm fixes the race
+// in the lock request retry algorihm and this test should no longer hang.
+
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Comment lines too long

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+namespace toku {
+
+// use 1000 when after_retry_all is implemented, otherwise use 100000
+static const int n_tests = 1000; // 100000;
+
+static void after_retry_all(void) {
+    usleep(10000);
+}
+
+static void run_locker(locktree *lt, TXNID txnid, const DBT *key, pthread_barrier_t *b) {
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Line too long

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+    for (int i = 0; i < n_tests; i++) {
+        int r;
+        r = pthread_barrier_wait(b); assert(r == 0 || r == PTHREAD_BARRIER_SERIAL_THREAD);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Line too long

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+
+        lock_request request;
+        request.create();
+
+        request.set(lt, txnid, key, key, lock_request::type::WRITE, false);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Line too long

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+            // retry pending lock requests
+            lock_request::retry_all_lock_requests(lt, after_retry_all);
+        }
+
+        request.destroy();
+        memset(&request, 0xab, sizeof request);
+
+        toku_pthread_yield();
+        if ((i % 10) == 0)
+            std::cout << std::this_thread::get_id() << " " << i << std::endl;
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Line too long, prefer not to use std::iostreams, prefer printf to stderr instead

File: locktree/tests/lock_request_start_retry_race_3.cc [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+    toku::locktree lt;
+    DICTIONARY_ID dict_id = { 1 };
+    lt.create(nullptr, dict_id, toku::dbt_comparator);
+
+    const DBT *one = toku::get_dbt(1);
+
+    const int n_workers = 3;
+    std::thread worker[n_workers];
+    pthread_barrier_t b;
+    int r = pthread_barrier_init(&b, nullptr, n_workers); assert(r == 0);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_race_3.cc]: newline

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 17]
```diff
@@ -0,0 +1,128 @@
...
+
+// Suppose that 2 threads are running a lock acquire, release, retry sequence.  There is a
+// race between the acquire and the release with 2 threads.  If thread 1 acquires a lock,
+// and thread 2 tries to acquire the same lock and fails, thread 1 may release its lock and retry
+// pending lock requests BEFORE thread 2 adds itself to the pending lock requests.  If this
+// happens, then thread 2 will HANG until its lock timer expires even when the lock it is
+// waiting for is FREE.
+
+// This test exposes this problem as a test hang.  If the race is fixed, then the test runs to
+// completion.
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Comment lines too long

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+
+namespace toku {
+
+static void start_before_pending(void) {
+    usleep(10000);
+}
+
+static void run_locker(locktree *lt, TXNID txnid, const DBT *key, pthread_barrier_t *b) {
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Line too long

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+    for (int i = 0; i < 100000; i++) {
+        int r;
+        r = pthread_barrier_wait(b); assert(r == 0 || r == PTHREAD_BARRIER_SERIAL_THREAD);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Line too long

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 20]
```diff
@@ -0,0 +1,128 @@
...
+
+        lock_request request;
+        request.create();
+        request.set(lt, txnid, key, key, lock_request::type::WRITE, false);
+
+        // if the callback is included, then the race is easy to reproduce.  Otherwise, several
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Line too long

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 21]
```diff
@@ -0,0 +1,128 @@
...
+        // test runs may be required before the race happens.
+        if (1) request.set_start_before_pending_test_callback(start_before_pending);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: if (1) ?
Line too long

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 22]
```diff
@@ -0,0 +1,128 @@
...
+            // retry pending lock requests
+            lock_request::retry_all_lock_requests(lt);
+        }
+
+        request.destroy();
+        memset(&request, 0xab, sizeof request);
+
+        toku_pthread_yield();
+        if ((i % 10) == 0)
+            std::cout << std::this_thread::get_id() << " " << i << std::endl;
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Line too long, prefer not to use std::iostreams, prefer printf to stderr instead

File: locktree/tests/lock_request_start_retry_wait_race_2.cc [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+    toku::locktree lt;
+    DICTIONARY_ID dict_id = { 1 };
+    lt.create(nullptr, dict_id, toku::dbt_comparator);
+
+    const DBT *one = toku::get_dbt(1);
+
+    const int n_workers = 2;
+    std::thread worker[n_workers];
+    pthread_barrier_t b;
+    int r = pthread_barrier_init(&b, nullptr, n_workers); assert(r == 0);
```
Comment (georgelorchpercona) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: newline

File: locktree/lock_request.cc (hunk 1 above)
Comment (vlad-lesin) [inline locktree/lock_request.cc]: Yep, you are right, I forgot to pass the patch through formatter. But even after autoformatting these lines stay the same, i.e. there is no line size limit violation here.

File: locktree/lock_request.cc (hunk 2 above)
Comment (vlad-lesin) [inline locktree/lock_request.cc]: Fixed.

File: locktree/lock_request.cc (hunk 3 above)
Comment (vlad-lesin) [inline locktree/lock_request.cc]: Fixed.

File: locktree/lock_request.cc (hunk 4 above)
Comment (vlad-lesin) [inline locktree/lock_request.cc]: There is no line size limit violation here.

File: locktree/lock_request.cc (hunk 5 above)
Comment (vlad-lesin) [inline locktree/lock_request.cc]: There is no line size limit violation here.

File: locktree/lock_request.h (hunk 6 above)
Comment (vlad-lesin) [inline locktree/lock_request.h]: Fixed.

File: locktree/locktree.cc (hunk 7 above)
Comment (vlad-lesin) [inline locktree/locktree.cc]: Fixed.

File: locktree/tests/lock_request_start_release_wait.cc (hunk 8 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_release_wait.cc]: Fixed.

File: locktree/tests/lock_request_start_release_wait.cc (hunk 9 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_release_wait.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_race.cc (hunk 10 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race.cc]: Why not to use std::iostream?  std::this_thread::get_id() returns std::thread::id class. This is not uint64_t. This class has an implementation of '<<' operator which serializes thread id. So if we are going to use fprintf() we have also to use something like std::stringstream to serialize thread id. If so then why not to use std::cerr?

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 11 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 12 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 13 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 14 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: There is no line size limit violation here.

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 15 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: "Line to long" fixed. Comment about std::iostream see above.

File: locktree/tests/lock_request_start_retry_race_3.cc (hunk 16 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_race_3.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 17 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 18 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 19 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 20 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 21 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 22 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: "Line too long" fixed.  See above the comment about std::iostream.

File: locktree/tests/lock_request_start_retry_wait_race_2.cc (hunk 23 above)
Comment (vlad-lesin) [inline locktree/tests/lock_request_start_retry_wait_race_2.cc]: Fixed.

**Comment by vlad-lesin** [comment]
George, I have used clang-format for this patch and manually checked all your notes.

File: locktree/lock_request.cc (hunk 4 above)
Comment (georgelorchpercona) [inline locktree/lock_request.cc]: I apologize, diff formatting in github made it appear too long.

**Comment by vlad-lesin** [comment]
The testing after the merging to the branch forked from the current master: http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1028/, http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1941/ .

File: locktree/lock_request.cc [hunk 24]
```diff
@@ -339,10 +377,6 @@ void lock_request::retry_all_lock_requests(locktree *lt) {
             i++;
         }
     }
-
-    // future threads should only retry lock requests if some still exist
-    info->should_retry_lock_requests = info->pending_lock_requests.size() > 0;
```
Comment (fungo) [inline locktree/lock_request.cc]: @vlad-lesin  this line be removed after merge, but still exists in the latest master, as we can see here:
https://github.com/percona/PerconaFT/blob/43f81569cfd002763e080689d720e52265b72206/locktree/lock_request.cc#L382-L384

**Pull request merged**